ttnn/cpp/ttnn/tensor/tensor_impl.cpp

```diff
@@ -809,6 +816,197 @@ Tensor to_device<bfloat8_b>(
     return to_device<uint32_t>(tensor, target_device, memory_config, cq_id, sub_device_ids);
 }
 
+// ======================================================================================
+//     Helpers for converting between logical <-> physical data with full tensor spec
+// ======================================================================================
+namespace CMAKE_UNIQUE_NAMESPACE {
+
+// TODO: Remove when we generalize interleaved and sharded; when we do, directly get from TensorLayout
+std::array<Size, 2> get_logical_and_physical_shard_shapes(const TensorSpec& tensor_spec) {
+    if (tensor_spec.memory_config().is_sharded()) {
+        return {
+            tensor_spec.tensor_layout().get_logical_shard_shape(),
+            tensor_spec.tensor_layout().get_physical_shard_shape()};
+    }
+
+    const auto& logical_shape = tensor_spec.logical_shape();
+    Size logical_shard_shape{logical_shape[-2], logical_shape[-1]};
+    auto physical_shard_shape = logical_shard_shape;
+    if (tensor_spec.layout() == Layout::TILE) {
+        const auto& tile = tensor_spec.tile();
+        auto physical_shard_height = tt::round_up(logical_shard_shape.height(), tile.get_height());
+        auto physical_shard_width = tt::round_up(logical_shard_shape.width(), tile.get_width());
+        physical_shard_shape = Size{physical_shard_height, physical_shard_width};
+    }
+    return {logical_shard_shape, physical_shard_shape};
+}
+
+using LogicalPhysicalIdxPairs = std::vector<std::pair<size_t, size_t>>;
+using LogicalPhysicalMapping = std::pair<LogicalPhysicalIdxPairs, size_t>;
+std::vector<LogicalPhysicalMapping> compute_logical_to_physical_shards_mapping(
+    const Size& logical_2D_shape,
+    const Size& logical_shard_shape,
+    const Size& physical_shard_shape,
+    const size_t physical_stride) {
+    const auto logical_stride = logical_2D_shape.width();
+
+    const auto shard_division_spec = tt::tt_metal::compute_shard_division_spec(logical_2D_shape, logical_shard_shape);
+    const auto num_shards_height = shard_division_spec.num_shards_height;
```

Comment:
Here too, I thhink you can decompose using `const auto [...] = shard_division_spec;`

Reply:
yep, it works